src/Ledgers.jl

```diff
@@ -12,94 +12,119 @@ Licensed under MIT License, see LICENSE.md
+    isdebit::Bool
 
-    function AccountInfo(::Type{T}, account, code, name) where {T<:AccountType}
-        return new{T}(account, code, name)
+    function AccountInfo{B}(account::Account{B}, code, name, isdebit=true, parent=nothing) where {B <: Position}
+        acc = new{B}(account, code, name, isdebit)
+        parent === nothing || push!(parent.subaccounts, acc)
+        acc
     end
 end
 
-struct AccountGroup{AT<:AccountType} <: AccountNode
+AccountInfo(account::Account{B}, code, name, isdebit=true, parent=nothing) where {B <: Position} =
+    AccountInfo{B}(account, code, name, isdebit, parent)
+
+struct AccountGroup{B <: Position} <: AccountNode{B}
+    id::AccountId
     code::AccountCode
     name::String
-    parent::Union{Nothing,AccountGroup{<:AccountType}}
-    subaccounts::Vector{AccountInfo}
-    subgroups::Vector{AccountGroup}
-
-    function AccountGroup(::Type{T}, account, name, parent=nothing) where {T<:AccountType}
-        new{T}(account, code, name, parent, Vector{AccountInfo}(), Vector{AccountGroup}())
+    isdebit::Bool
+    parent::Union{Nothing,AccountGroup{B}}
+    subaccounts::StructArray{AccountInfo{B}}
+    subgroups::StructArray{AccountGroup{B}}
+
+    function AccountGroup{B}(
+        id,
+        code,
+        name,
+        isdebit=true,
+        parent=nothing,
+        subaccounts=StructArray(Vector{AccountInfo{B}}()),
+        subgroups=StructArray(Vector{AccountGroup{B}}())) where {B <: Position}
+        acc = new{B}(id, code, name, isdebit, parent, subaccounts, subgroups)
+        parent === nothing || push!(parent.subgroups, acc)
+        acc
     end
 end
 
+AccountGroup(
+    ::B,
+    code,
+    name,
+    isdebit=true,
+    parent=nothing, 
+    subaccounts=StructArray(Vector{AccountInfo{B}}()),
+    subgroups=StructArray(Vector{AccountGroup{B}}())) where {B <: Position} =
+    AccountGroup{B}(AccountId(), code, name, isdebit, parent, subaccounts, subgroups)
+
 # Identity function (to make code more generic)
-account(acc::Account) = acc
+account(acc::AccountType) = acc
 account(info::AccountInfo) = info.account
 
-account_type(::Union{AccountInfo{AT},AccountGroup{AT}}) where {AT} = AT
-code(acc::Union{<:AccountInfo,<:AccountGroup}) = acc.code
-name(acc::Union{<:AccountInfo,<:AccountGroup}) = acc.name
+code(acc::AccountNode) = acc.code
+name(acc::AccountNode) = acc.name
+isdebit(acc::AccountNode) = acc.isdebit
 
 parent(group::AccountGroup) = group.parent
 subaccounts(group::AccountGroup) = group.subaccounts
 subgroups(group::AccountGroup) = group.subgroups
 
-id(acc::Account) = acc.id
-id(info::AccountInfo) = id(account(info))
+id(acc::AccountType) = account(acc).id
 
-balance(acc::Account) = acc.balance
-balance(info::AccountInfo) = balance(account(info))
-balance(group::AccountGroup) = 
-    sum(map(info->balance(info), subaccounts(group))) + 
-    sum(map(grp->balance(grp), subgroups(group)))
+balance(acc) = account(acc).balance
+balance(group::AccountGroup) = isempty(subgroups(group)) ?
```

Comment:
Something more like this, I think:
```
_sum_vec(vec, T) = isempty(vec) ? zero(T) : sum(balance, vec)
balance(acc::AccountGroup{P}) where {S,T,P<:Position{S,T}} =
    _sum_vec(subgroups(acc), T) + _sumvec(subaccounts(acc), T)
```

Reply:
Yeah. I am a little confused by this. If `subaccounts` is empty, the sum is `I(0)` (where `I` is some instrument type) as it should be, but if `subgroups` is empty, it fails. I'd rather figure out why `subgroups` is failing than add a new internal method like that.

Reply:
It might be due to the recursive nature of that.